File: lib/to_mdx/parsers/open_document.rb
# frozen_string_literal: true

require "zip"
require "fileutils"
require "date"
require "tmpdir"
require "optparse"

require_relative "./open_document/base"
require_relative "./open_document/all_parsers"
require_relative "../styles"
require_relative "../generators/mdx"
require_relative "../page_dimensions"

module ToMdx
  module Parsers
    module OpenDocument
      class << self
        def parse(output_root, presentation_file, options = {})
          Zip::File.open(presentation_file) do |zip_file|
            content_doc = nil
            meta_doc = nil
            styles_doc = nil

            zip_file.each do |entry|
              case entry.name
              when "content.xml"
                content_doc = ToMdx::Parsers::Base.parse_xml(entry.get_input_stream.read)
              when "meta.xml"
                meta_doc = ToMdx::Parsers::Base.parse_xml(entry.get_input_stream.read)
              when "styles.xml"
                styles_doc = ToMdx::Parsers::Base.parse_xml(entry.get_input_stream.read)
              end
            end

            unless content_doc && meta_doc && styles_doc
              raise Error("Error: Could not find content.xml or meta.xml in the presentation file.")
            end

            page_dimensions = ToMdx::PageDimensions.infer(content_doc, styles_doc)
            if page_dimensions
              puts("🔍 Found page dimensions: #{page_dimensions[:width]}cm x #{page_dimensions[:height]}cm")
            end

            # Create empty styles to avoid circular dependency issues
            styles = ToMdx::Styles.new({})

            metadata = {}
            parsed_meta_doc = ToMdx::Parsers::Base.parse(meta_doc.root, metadata)

            title = get_presentation_title(metadata)
            subtitle = metadata["subtitle"]
            description = metadata["description"]
            date = get_presentation_date(metadata)
            basename = "#{date.strftime('%Y-%m-%d')}_#{sanitize_filename(title)}"

            context = {
              metadata: metadata,
              basename: basename,
              options: options,
              page_dimensions: page_dimensions,
              styles: styles,
            }

            extract_images(zip_file, basename, output_root)

            frontmatter = ToMdx::Generators::Mdx.generate_frontmatter(metadata)
            content = ToMdx::Parsers::Base.parse(content_doc.root, context)

            relative_output_file = File.join("src/pages/presentations", generate_filename(date, title))
            output_file = File.join(output_root, relative_output_file)
            FileUtils.mkdir_p(File.dirname(output_file))

            File.open(output_file, "w") do |f|
              f.write(frontmatter)
              f.write("\n")
              f.write(content.to_mdx)
              f.write("\n")
            end

            puts("")
            puts("✅ #{relative_output_file}")
          end
        end

        private

        def get_presentation_title(metadata)
          title = metadata["title"]
          if title && !title.empty?
            puts("🔍 Found presentation title: #{title}")
            return title
          end

          print("Enter presentation title: ")
          $stdout.flush
          STDIN.gets.chomp
        end

        def get_presentation_date(metadata)
          date = metadata["date"]
          if date
            puts("🔍 Found presentation date: #{date}")
            return date
          end

          print("Enter presentation date (YYYY-MM-DD): ")
          $stdout.flush
          date_input = STDIN.gets.chomp

          begin
            Date.parse(date_input)
          rescue Date::Error
            puts("Invalid date format. Please use YYYY-MM-DD.")
            exit(1)
          end
        end

        def sanitize_filename(text)
          text.downcase
              .gsub(/[^a-z0-9\s]/, "")
              .gsub(/\s+/, "_")
        end

        def generate_filename(date, title)
          basename = "#{date.strftime('%Y-%m-%d')}_#{sanitize_filename(title)}"
          "#{basename}.mdx"
        end

        def extract_images(zip_file, basename, project_root)
          images_dir = File.join(project_root, "src", "static", "img", "presentations", basename)
          FileUtils.mkdir_p(images_dir)

          zip_file.each do |entry|
            next unless entry.name.start_with?("Pictures/")

            target_path = File.join(images_dir, entry.name.sub(/^Pictures\//, ""))

            begin
              entry.extract(target_path)
            rescue Zip::DestinationFileExistsError
              # Skip if file already exists
            end
          end
        end
      end
    end
  end
end

File: src/parsers/open_document.ts
/* eslint-disable @typescript-eslint/prefer-for-of */
import * as fs from "fs";
import * as path from "path";
import * as yauzl from "yauzl";
import { formatDate, generateFilename, generateMetadataExports, sanitizeFilename, titleFromPath } from "../generators/mdx.ts";
import { infer } from "../page-dimensions.ts";
import type { Options } from "../parsers.ts";
import { Styles } from "../styles.ts";
import type { ParseContext } from "./base_element.ts";
import { BaseElement } from "./base_element.ts";

import dedent from "dedent-js";
import "./open-document/all_parsers.ts"; // Register all of the parsers

export async function parse(outputRoot: string, presentationFile: string, options: Options = {}): Promise<void> {
  let contentDoc: string | null = null;
  let metaDoc: string | null = null;
  let stylesDoc: string | null = null;

  const docs = await extractXmlDocuments(presentationFile);
  contentDoc = docs.content;
  metaDoc = docs.meta;
  stylesDoc = docs.styles;

  if (!contentDoc || !metaDoc || !stylesDoc) {
    throw new Error("Error: Could not find content.xml, meta.xml, or styles.xml in the presentation file.");
  }

  const contentDocument = BaseElement.parseXml(contentDoc);
  const metaDocument = BaseElement.parseXml(metaDoc);
  const stylesDocument = BaseElement.parseXml(stylesDoc);

  const pageDimensions = infer(contentDocument, stylesDocument);
  if (pageDimensions) {
    console.log(`🔍 Found page dimensions: ${pageDimensions.width}cm x ${pageDimensions.height}cm`);
  }

  // TODO avoid this cast, but find a way so that styles is always in the context
  const metadata: Record<string, unknown> = {};
  BaseElement.parse(metaDocument.documentElement, metadata as unknown as ParseContext);

  const title = await getPresentationTitle(metadata, presentationFile);
  const date = await getPresentationDate(metadata, presentationFile);
  const basename = `${formatDate(date)}_${sanitizeFilename(title)}`;

  const context: ParseContext = {
    metadata,
    basename,
    options,
    pageDimensions: pageDimensions || undefined,
    styles: new Styles(),
  };

  await extractImages(presentationFile, basename, outputRoot);

  const metadataExports = generateMetadataExports(metadata);

  BaseElement.parse(stylesDocument.documentElement, context);
  const content = BaseElement.parse(contentDocument.documentElement, context);

  const relativeOutputFile = path.join("src/pages/presentations", generateFilename(date, title));
  const outputFile = path.join(outputRoot, relativeOutputFile);

  // Ensure directory exists
  const outputDir = path.dirname(outputFile);
  if (!fs.existsSync(outputDir)) {
    fs.mkdirSync(outputDir, { recursive: true });
  }

  fs.writeFileSync(
    outputFile,
    dedent`
      ${metadataExports}
      ${content?.toMdx() ?? ""}
    `,
  );

  console.log("");
  console.log(`✅ ${relativeOutputFile}`);
}

async function extractXmlDocuments(presentationFile: string): Promise<{
  content: string | null;
  meta: string | null;
  styles: string | null;
}> {
  const docs = { content: null as string | null, meta: null as string | null, styles: null as string | null };

  return await new Promise((resolve, reject) => {
    yauzl.open(presentationFile, { lazyEntries: true }, (err, zipFile) => {
      if (err) {
        reject(err);
        return;
      }

      zipFile.on("entry", (entry: yauzl.Entry) => {
        if (entry.fileName !== "content.xml" && entry.fileName !== "meta.xml" && entry.fileName !== "styles.xml") {
          zipFile.readEntry();
          return;
        }

        zipFile.openReadStream(entry, (err, readStream) => {
          if (err) {
            reject(err);
            return;
          }

          let content = "";
          readStream.on("data", (chunk: Buffer) => {
            content += chunk.toString();
          });

          readStream.once("end", () => {
            if (entry.fileName === "content.xml") docs.content = content;
            else if (entry.fileName === "meta.xml") docs.meta = content;
            else if (entry.fileName === "styles.xml") docs.styles = content;
            zipFile.readEntry();
          });

          readStream.on("error", reject);
        });
      });

      zipFile.once("error", reject);
      zipFile.once("end", () => {
        resolve(docs);
      });

      zipFile.readEntry();
    });
  });
}

async function extractImages(zipFileName: string, basename: string, projectRoot: string): Promise<void> {
  const imagesDir = path.join(projectRoot, "src/static/img/presentations", basename);
  if (!fs.existsSync(imagesDir)) {
    fs.mkdirSync(imagesDir, { recursive: true });
  }

  return new Promise((resolve, reject) => {
    yauzl.open(zipFileName, { lazyEntries: true }, (err, zipFile) => {
      if (err) {
        reject(err);
        return;
      }

      zipFile.on("entry", (entry: yauzl.Entry) => {
        if (!entry.fileName.startsWith("Pictures/") || entry.fileName.endsWith(".svm")) {
          zipFile.readEntry();
          return;
        }

        const filename = entry.fileName.replace(/^Pictures\//, "");
        const targetPath = path.join(imagesDir, filename);
        if (fs.existsSync(targetPath)) {
          zipFile.readEntry();
          return;
        }

        zipFile.openReadStream(entry, (streamErr, readStream) => {
          if (streamErr) {
            reject(streamErr);
            return;
          }

          const writeStream = fs.createWriteStream(targetPath);
          readStream.pipe(writeStream);

          writeStream.once("error", reject);
          writeStream.once("close", () => {
            zipFile.readEntry();
            resolve();
          });
        });
      });

      zipFile.on("end", () => {
        resolve();
      });

      zipFile.on("error", reject);
      zipFile.readEntry();
    });
  });
}

async function getPresentationTitle(metadata: Record<string, unknown>, presentationFile: string): Promise<string> {
  const title = metadata.title;
  if (typeof title === "string" && title.trim()) {
    console.log(`🔍 Found presentation title: ${title}`);
    return title;
  }

  // Degrade to the file name rather than throwing (mirrors the Keynote path).
  const fallback = titleFromPath(presentationFile);
  console.warn(`⚠️  No presentation title in metadata; using file name "${fallback}"`);
  return fallback;
}

async function getPresentationDate(metadata: Record<string, unknown>, presentationFile: string): Promise<Date> {
  const presentationDate = metadata.presentation_date;
  if (typeof presentationDate === "string") {
    const parsedDate = new Date(presentationDate);
    if (!isNaN(parsedDate.getTime())) {
      console.log(`🔍 Found presentation date: ${presentationDate}`);
      return parsedDate;
    }
  }

  const date = metadata.date;
  if (date instanceof Date) {
    console.log(`🔍 Found presentation date: ${date.toISOString().split("T")[0]}`);
    return date;
  }

  if (typeof date === "string") {
    const parsedDate = new Date(date);
    if (!isNaN(parsedDate.getTime())) {
      console.log(`🔍 Found presentation date: ${date}`);
      return parsedDate;
    }
  }

  // Degrade to the file's creation time rather than throwing (mirrors the Keynote
  // path). birthtime is unset on some filesystems, where it reads as epoch 0 — fall
  // back to mtime there.
  const stats = fs.statSync(presentationFile);
  const created = stats.birthtimeMs > 0 ? stats.birthtime : stats.mtime;
  console.warn(`⚠️  No presentation date in metadata; using file creation time ${formatDate(created)}`);
  return created;
}
